src/tasks/withdraw.ts

```diff
@@ -173,12 +173,56 @@ async function appraise(
   return { ...token, usdValue, decimals };
 }
 
-async function getAllTradedTokens(settlement: Contract): Promise<string[]> {
-  const trades = await settlement.queryFilter(settlement.filters.Trade());
-  const tokens = new Set(
-    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-    trades.map((trade) => [trade.args!.sellToken, trade.args!.buyToken]).flat(),
-  );
+function isErrorTooManyEvents(error: Error): boolean {
+  return /query returned more than \d* results/.test(error.message);
+}
+
+// List all traded tokens. Block range bounds are both inclusive.
+async function getAllTradedTokens(
+  settlement: Contract,
+  fromBlock: number,
+  toBlock: number,
+  hre: HardhatRuntimeEnvironment,
+): Promise<string[]> {
+  let trades = null;
+  try {
+    trades = await hre.ethers.provider.getLogs({
+      topics: [settlement.interface.getEventTopic("Trade")],
+      fromBlock,
+      toBlock,
+    });
+  } catch (error) {
+    if (!isErrorTooManyEvents(error)) {
+      throw error;
+    }
+  }
+
+  let tokens;
+  if (trades === null) {
+    if (fromBlock === toBlock) {
+      throw new Error("Too many events in the same block");
+    }
```

Comment:
So if we actually run into this eventually, doesn't this error essentially mean the script (as is) can not complete and we would have to adapt it again?

Reply:
Yes. On the other hand, the current event limit is 10k, which means that assuming the block gets filled then each event would have at its disposal 1500 gas, which is quite less than the gas needed to settle a trade, so I don't expect this to ever happen.
It's theoretically possible in general, however, since I think that emitting an event costs less that that. Maybe this Friday I'll try to fill a Rinkeby block with events and see what happens. :smile:

Reply:
The largest amount of events _with a topic_ (so a standard Solidity event) that I could squeeze into a Rinkeby block is 12535 ([tx](https://rinkeby.etherscan.io/tx/0x028da9a789d80c7fb6403ca7a07d0c461ee66db57f18ce831874a34f2add847f)), and I was able to retrieve all events with the same call that I made in this PR. This is only true if my range is a single block, If I include another empty block in the range of the same query then it returns an error, so probably Infura allows you to query more than the maximum number of events in a single block. On mainnet, I should be able to fit 18803 events.
By the way, the transaction took 1733.76s to be mined and I wasn't able to test it with Ganache because it kept crashing. :sweat_smile:

Reply:
I've beat my personal record on a [single transaction](https://rinkeby.etherscan.io/tx/0x0c524f0ff1f249e479e1911827285e7e689452326acc739a9ba1fcbfaca4db21) with 39472 events. The current overhead for each event is about 4.5 gas/event. :tada: